aws-lc-rs/src/key_transport.rs

```diff
@@ -0,0 +1,389 @@
+    /// from the operation and then returns what `kdf` returns.
+    /// 
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    /// 
+    pub fn decapsulate<F, R>(&self, ciphertext: &mut [u8], kdf: F) -> Result<R, Unspecified>
+    where
+        F: FnOnce(&[u8]) -> Result<R, Unspecified> {
+        unsafe {
+            let ctx = DetachableLcPtr::new(EVP_PKEY_CTX_new(*self.context, null_mut()))?;
+            let mut shared_secret_len;
+            match self.algorithm {
+                KemAlgorithm::KYBER512_R3 => {
+                    shared_secret_len = KYBER512_BYTES;
+                }
+            }
+            let mut shared_secret: Vec<u8> = vec![0u8; shared_secret_len];
+            if EVP_PKEY_decapsulate(*ctx, shared_secret.as_mut_ptr(), &mut shared_secret_len,
+                                    ciphertext.as_mut_ptr(), ciphertext.len()) != 1 {
+                shared_secret.zeroize()
+            }
+            kdf(&shared_secret)
+        }
+    }
+
+    /// Creates a new KEM private key from raw bytes
+    /// 
+    /// `alg` is the `KemAlgorithm` to be associated with the generated `KemPrivateKey`
+    /// 
+    /// `bytes` is a slice of raw bytes representing a `KemPrivateKey`
+    /// 
+    /// # Errors
+    /// `error::KeyRejected` when operation fails during key creation.
+    /// 
+    pub fn from_raw_bytes(alg: KemAlgorithm, bytes: &[u8]) -> Result<Self, KeyRejected> {
+        unsafe {
+            let pkey = DetachableLcPtr::new(EVP_PKEY_kem_new_raw_secret_key(alg.nid(), bytes.as_ptr(), bytes.len())).map_err(|_e| KeyRejected::unexpected_error())?;
+            Ok(KemPrivateKey {
+                algorithm: alg,
+                context: LcPtr::from(pkey),
+                priv_key: bytes.to_owned()
+            })
+        }
+    }
+
+    /// Gets the raw bytes of the current `KemPrivateKey`
+    pub fn get_raw_bytes(&self) -> &[u8] {
+        &self.priv_key
+    }
+}
+
+#[cfg(test)]
+impl PartialEq for KemPrivateKey {
+    fn eq(&self, other: &Self) -> bool {
+        let mut self_privkey_bytes_from_ctx = vec![0u8; self.algorithm.get_privkey_len()];
+        let mut other_privkey_bytes_from_ctx = vec![0u8; other.algorithm.get_privkey_len()];
+
+        unsafe {
+            if 1 != EVP_PKEY_get_raw_private_key(*self.context, self_privkey_bytes_from_ctx.as_mut_ptr(), &mut self.algorithm.get_privkey_len()) {
+                panic!("Error getting private key from context");
+            }
+            if 1 != EVP_PKEY_get_raw_private_key(*other.context, other_privkey_bytes_from_ctx.as_mut_ptr(), &mut other.algorithm.get_privkey_len()) {
+                panic!("Error getting private key from context");
+            }
+        }
+        self.algorithm == other.algorithm &&
+        self.priv_key == other.priv_key &&
+        self_privkey_bytes_from_ctx == other_privkey_bytes_from_ctx
+    }
```

Comment:
I suppose, yes. The reason I added this was to essentially check the validity of the EVP_PKEY pointer. Since when reading from raw bytes technically these aren't exactly the same, as the raw bytes are written directly to the buffer while the EVP_PKEY is constructed from them.
```rust
let pkey = DetachableLcPtr::new(EVP_PKEY_kem_new_raw_secret_key(
    alg.nid(),
    bytes.as_ptr(),
    bytes.len(),
))?;
Ok(KemPrivateKey {
    algorithm: alg,
    context: LcPtr::from(pkey),
    priv_key: bytes.to_owned(),
})
``` 

If this is unnecessary I can remove this extra check.